### PR TITLE
Support forward mode differentiation for SVI

### PR DESCRIPTION
**Pull request by juanitorduz**
Closes https://github.com/pyro-ppl/numpyro/issues/1726

Trying here a "good first issue".

### CODE REVIEW

**Comment by fehiepsi** [comment]
Looks great to me. Could you add a simple test rwith while loop in the model?

**Comment by juanitorduz** [comment]
I added a simple test `test/infer/test_svi.py::test_forward_mode_differentiation` in https://github.com/pyro-ppl/numpyro/pull/1731/commits/532fb5b186d445fda626c3eb682a3d8b666dfb21 but it is failing with

```python
    @functools.wraps(update)
    def tree_update(i, grad_tree, opt_state):
      states_flat, tree, subtrees = opt_state
      grad_flat, tree2 = tree_flatten(grad_tree)
      if tree2 != tree:
        msg = ("optimizer update function was passed a gradient tree that did "
               "not match the parameter tree structure with which it was "
               "initialized: parameter tree {} and grad tree {}.")
>       raise TypeError(msg.format(tree, tree2))
E       TypeError: optimizer update function was passed a gradient tree that did not match the parameter tree structure with which it was initialized: parameter tree PyTreeDef({'loc': *, 'scale': *}) and grad tree PyTreeDef(({'loc': *, 'scale': *}, None))
```

And I am not sure if its because of the implementation of the test is wrong 😑 . Any tips? Thanks

File: numpyro/optim.py [hunk 1]
```diff
@@ -34,6 +34,11 @@
 _OptState = TypeVar("_OptState")
 _IterOptState = tuple[int, _OptState]
 
+def _value_and_grad(f, x, forward_mode_differentiation=False):
+    if forward_mode_differentiation:
+        return f(x), jacfwd(f, has_aux=True)(x)
```
Comment (fehiepsi) [inline numpyro/optim.py:39]: I think you can set has_aux to False.

A faster way, which just does 1 forward pass, is to redefine f:
```
def wrapper(x):
  out, aux = f(x)
  return out, (out, aux)

grad, (out, aux) = jacfwd(f, has_aux=True)(x)
return (out, aux), grad
```

We can also apply this trick in hmc forward mode implementation.

Comment (juanitorduz) [reply numpyro/optim.py]: ok! So I added  your suggestion in https://github.com/pyro-ppl/numpyro/pull/1731/commits/ef3a3f76ba53fe984b50f4adfb410c71913d3b4b and now I getting and error downstream 🤔 . Do I need to further unpack somewhere?

```python
        if progress_bar:
            losses = []
            with tqdm.trange(1, num_steps + 1) as t:
                batch = max(num_steps // 20, 1)
                for i in t:
                    svi_state, loss = jit(body_fn)(svi_state, None)
                    losses.append(jax.device_get(loss))
                    if i % batch == 0:
                        if stable_update:
                            valid_losses = [x for x in losses[i - batch :] if x == x]
                            num_valid = len(valid_losses)
                            if num_valid == 0:
                                avg_loss = float("nan")
                            else:
                                avg_loss = sum(valid_losses) / num_valid
                        else:
>                           avg_loss = sum(losses[i - batch :]) / batch
E                           TypeError: unsupported operand type(s) for +: 'int' and 'dict'

numpyro/infer/svi.py:409: TypeError
```

Comment (juanitorduz) [reply numpyro/optim.py]: It worked in https://github.com/pyro-ppl/numpyro/pull/1731/commits/a64c092f2e630b99fc4833927e05292775519251 !

Comment (juanitorduz) [reply numpyro/optim.py]: Same trick in https://github.com/pyro-ppl/numpyro/pull/1731/commits/1259db7b79ac1aecce97fa87a984200906c50e1e for hmc

**Review by fehiepsi — APPROVED** [review]
Woohoo, thanks for supporting this feature!

**Comment by juanitorduz** [comment]
Thank you for your guidance @fehiepsi 🙏🙂

**Pull request merged**